File: Datasructure/StackAndQueses/stack.js
class Node{
    constructor(value){
        this.value=value;
        this.next=null;
    }
}


class Stack {
    constructor(){
        this.top=null;
        this.bottom=null;
        this.length=0;
    }

    peek(){
        return this.top;
    }

    push(value){
        let newNode=new Node(value)
        if(this.length===0){
            this.top=newNode
            this.bottom=newNode
        }
        else{
            let holdingPointer=this.top
            this.top=newNode
            this.top.next=holdingPointer
        }
        this.length++;
        return this;

    }

    pop(){
        if(!this.top){
            return "No element in stack";
        }
        if(this.top===this.bottom){
            this.bottom=null;
        }
        const holdingPointer=this.top;
        this.top=this.top.next;
        this.length--;
        return holdingPointer;
    }

    isEmpty(){
        return this.length==0
    }
}
const myStack=new Stack()

console.log(myStack.peek())
console.log("------------------")
myStack.push('Dragon')
console.log(myStack.peek())
console.log(myStack.push('Human').peek())
console.log(myStack.push('Angel').peek())
console.log(myStack.pop())
console.log(myStack.pop())
console.log(myStack.peek())

File: Datasructure/StackAndQueses/stack.py
class NewNode:

    def __init__(self, value: int):
        self.value = value
        self.next = None


class MyStack:

    def __init__(self):
        self.topVal = None
        self.bottom = None
        self.length = 0

    def push(self, x: int) -> None:
        newNode = NewNode(x)
        if(self.length == 0):
            self.button = newNode
            self.topVal = newNode
        else:
            pointer = self.topVal
            self.topVal = newNode
            self.topVal.next = pointer
        self.length += 1

    def pop(self) -> int:
        if(self.topVal == None):
            return -1
        if(self.topVal == self.bottom):
            self.button = None
        pointer = self.topVal
        self.topVal = self.topVal.next
        self.length -= 1
        return pointer.value

    def top(self) -> int:
        return self.topVal.value

    def empty(self) -> bool:
        return(self.length == 0)


# Your MyStack object will be instantiated and called as such:
obj = MyStack()
obj.push(3)
param_2 = obj.pop()
param_3 = obj.top()
param_4 = obj.empty()
print(param_2, param_3, param_4)
